Fix some unsafe accesses to OCaml values

braibant opened the pull request:
The current conversion from Rust values to OCaml values (represented by the instances of the trait `ToValue`) does not seem correct for nested values. For instance, consider the pair case:
```
let v = unsafe { ocaml_sys::caml_alloc_tuple(2) };
let res = pin(v);
T1::to_value(v1, |x| unsafe { ocaml_sys::store_field(v, 0, x) });
T2::to_value(v2, |x| unsafe { ocaml_sys::store_field(v, 1, x) });
res
```
It might be the case that converting `v2`  to an OCaml value allocates, and moves `v1` around. I don't think there is currently a guarantee that the value we stored in `v[0]` is still valid. I believe that the first commit in this branch segfaults throught exhibiting a variation of this issue.

This MR tries to implement `ToValue` using continuations (by first building a `RootedValue` to have a stable way to refer to OCaml values accross GCs, calling the provided continuation once the value has been assembled, then dropping the rooted value). 

Unfortunately, this branch does not fix the segfault at this stage.

braibant committed:
Fix some GC issues in ToValue

The trait `ToValue` is simplified, and instances are required to
implement a single function `to_value` allocating  an ocaml value (of
type `ocaml_sys::Value`) for a given `T`. It's the responsability of the
caller of `to_value` to either root this value or make it reachable from
a rooted ocaml value before the next allocation point.

LaurentMazare commented:
I'm not sure the example you mentioned is actually problematic, if `v1` is moved by the gc, as it has already been stored in the field 0 of v, the gc will update this reference which is fine. That said, it can be the case that the gc moves `v` when creating `v2` which would result in the pinning writing it to an incorrect address (this would only happen if `v` is not the *top level* value which has been boxrooted but instead an intermediary value below this).

Anyway, there is certainly an issue to be fixed in the current implementation. Is there any reason for keeping the continuation style here? This was mostly introduced because of the `pin` hack but if this does not work, maybe this could be entirely removed. E.g. the pair implementation would be something like:
```rust
impl<T1, T2> ToValue for (T1, T2)
where
    T1: ToValue,
    T2: ToValue,
{
    fn to_value(&self, pin: F) -> ocaml_sys::value 
    {
        let (v1, v2) = self;
        let v = unsafe { ocaml_sys::caml_alloc_tuple(2) };
        let rooted = rooted::RootedValue(v);
        ocaml_sys::store_field(v, 0, T1::to_value(v1));
        ocaml_sys::store_field(v, 1, T2::to_value(v2));
        drop(rooted);
        v
    }
}
```
If the goal is to make it harder to mess things up by running a function that could allocate after having dropping `rooted` and before `v` is actually stored somewhere. Maybe a better way around this would be to encode it in the Rust type system via something along the lines of [Safely Mixing OCaml and Rust](https://docs.google.com/viewer?a=v&pid=sites&srcid=ZGVmYXVsdGRvbWFpbnxtbHdvcmtzaG9wcGV8Z3g6NDNmNDlmNTcxMDk1YTRmNg) ?

(also note that the segfault related to custom blocks as I think independent of this and should hopefully be gone at the tip of the repo)

braibant committed:
Run Gc.compact() when allocating to showcase GC safety issues

braibant commented:
Hey. Sorry about that, my example write up was indeed wrong, but I think we are on the same page regarding the issue. I also think I have now pushed a better example, which relies on adding some GC stubs to `ocaml-rs` to expose a reliable way to trigger a compaction while allocating values from Rust (I need to clean this up and create a MR for this in the `ocaml-rs` project).

Looking at how to fix this, I am not sure I follow your implementation of `ToValue` for pairs here. (In the snipped below, I have modified your code a tiny bit, for the sake of the exposure, but I don't think I have changed the semantics). 

```rust
impl<T1, T2> ToValue for (T1, T2)
where
    T1: ToValue,
    T2: ToValue,
{
    fn to_value(&self) -> ocaml_sys::value 
    {
        let (v1, v2) = self;
        let v = unsafe { ocaml_sys::caml_alloc_tuple(2) };
        let rooted = rooted::RootedValue(v);
        let x1 = T1::to_value(v1);
        ocaml_sys::store_field(v, 0, x1); // Is `v` still valid here?
        let x2 T2::to_value(v2);
        ocaml_sys::store_field(v, 1, x2); // Is `v` still valid here?
        drop(rooted); 
        v  
    }
}
```

My mental model of this is that if the GC moves objects as part of the call to `T1::to_value(v1)`, the use of `v` in the `store_fields` calls is incorrect . I think one would need to use `rooted.value().value` instead (but even there, I think using intermediate let definitions to control evaluation order is necessary).

Re. continuation passing style vs. encoding things using the type system, I don't have a strong opinion, and need to look at this further. However, my initial reaction when looking at the pair example is that the caller would have to sometimes re-rootbox the returned value `v`, while using continuations, one would boxroot `v` exactly once, and drop the root once the continuation returns. The latter feels more safe and more efficient, but I am probably missing something.

LaurentMazare commented:
Ah indeed you're right, using `v` in the `store_field` would be unsafe as the gc could move things, but using `rooted.value()` instead should be fine. Also agreed that the order of evaluation of the parameters has to be fixed so the `let` makes this explicit. So it would become something like:
```rust
impl<T1, T2> ToValue for (T1, T2)
where
    T1: ToValue,
    T2: ToValue,
{
    fn to_value(&self) -> ocaml_sys::value 
    {
        let (v1, v2) = self;
        let v = unsafe { ocaml_sys::caml_alloc_tuple(2) };
        let rooted = rooted::RootedValue(v);
        let x1 = T1::to_value(v1);
        ocaml_sys::store_field(rooted.value(), 0, x1);
        let x2 = T2::to_value(v2);
        ocaml_sys::store_field(rooted.value(), 1, x2);
        rooted.value()  
    }
}
```
Re using the type system, `src/gc.rs` was an early attempt at this where holding a mut reference to the `gc::Gc` object is required to perform an operation that could allocate. The `ocaml_sys::value` type is abstracted behind a `Value<'a, T>` type for an OCaml object of type `T` with a lifetime used to track the dependency to `gc::Gc`. And creating a `rooted::RootedValue` is a way to remove the lifetime annotation from a value (and then calling `myrooted.value()` introduce a fresh new lifetime).
Not sure it's the right way to do it, one trickiness is how do you ensure that there is only a single `gc::Gc` object that can be created for example. Also as all this code is mostly to be used behind some `#[ocaml_rust::bridge]` macro, adding such static checks may not be necessary.

braibant commented:
I have updated this branch, to have `to_value` simply return an `ocaml_sys::Value` rather than use CPS. It means that the only thing that the caller of `to_value` can now do with returned values is either boxroot them immediately, or make them reachable from on OCaml root (e.g. by setting a field of a tuple which is itself reachable from an OCaml root). 

I was initially worried about the potential (efficiency) downside of having a `RootedValue` be dropped in an given instance of `to_value` (when `to_value` returns), to have the caller of `to_value` immediately have to re-root the said value. Upon inspection of the code I wrote in `to_value`, this is never the case (it's always possible to make the value reachable through a call to `store_field`).

LaurentMazare commented:
Looks good, I've left a comment for what I think might be an issue. It looks like the type system would be really helpful to catch these (especially for the crate users writing their own conversion functions), but even without that this PR is well worth it and a step in the good direction so that's probably better done in a separate PR.

braibant committed:
Fix some unsafe access to values

braibant commented on src/closure.rs:
```diff
@@ -148,9 +154,9 @@ where
-        let arg1 = arg1.to_value(|x| x);
-        let arg2 = arg2.to_value(|x| x);
-        let arg3 = arg3.to_value(|x| x);
+        let arg1 = arg1.to_value();
```
You're right. I double checked closure, and found a couple of other instances. Hopefully, this is better now.

LaurentMazare commented:
Merged, thanks for the PR!
